### PR TITLE
Introduce Device:useDPadAsActionKeys()

### LINKED ISSUE #11887: Resolve Device:hasFiveWay() code debt
* KOReader version: current master

#### Issue
It doesn't describe what it does. It's confusing and badly named.

https://github.com/koreader/koreader/pull/11884#discussion_r1614524106

> To summarize, imho when reading the text `hasFiveWay` it implies one thing and one thing only: that you can click/press on the d-pad or stick itself as a "fifth" action.[1] This is almost entirely useless information, except potentially to see it as a kind of restriction: maybe it's difficult to press on it without accidentally triggering a direction. For all intents and purposes it simply means we have four directions and an a primary action key.[2]
> 
> Instead it's being used as a secret code for two or three other features: that page turning buttons exist, and that some sufficient selection of other additional buttons exist so that actions like hold can be triggered.

### PR DESCRIPTION
**Pull request by Frenzie**
1. Non-Kindle-specific `hasFiveWay` behavior is changed to `hasDPad and useDPadAsActionKeys`. For now they remain Kindle-specific in practice, unless one sets `useDPadAsActionKeys = yes` in a user patch.
2. With that disambiguation out of the way, `hasFiveWay` itself is further disambiguated into `hasScreenKB` and `hasSymKey`, as per the actual property being used, rather than something that tends to correlate with it. (It needn't be Kindle-specific per se, but non-Kindle devices have equivalent shortcuts with for example `Shift`.)
  Running the emulator with `DISABLE_TOUCH=1` will set `hasSymKey = yes`, which can be tested with right shift.

Closes #11887.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11900)

### CODE REVIEW

**Comment by Frenzie** [comment]
@Commodore64user I think this should do it.

It's been reduced to two changes:

1. Non-Kindle-specific `hasFiveWay` behavior is changed to `hasDPad and useDPadAsActionKeys`. For now they remain Kindle-specific in practice, unless one sets `useDPadAsActionKeys = yes` in a user patch.
2. With that disambiguation out of the way, `hasFiveWay` is renamed to Kindle-specific `hasScreenKB`, as per the actual property being used, rather than something that tends to correlate with it. (It needn't be Kindle-specific per se, but non-Kindle devices have equivalent shortcuts with for example `Shift`.)

**Comment by Commodore64user** [comment]
Technically speaking, only K4 has a ScreenKB key, so it might be wise to use something else to dub isKindle with.

**Comment by Frenzie** [comment]
useKindleKeys?

**Comment by Commodore64user** [comment]
I think we could use `hasSymKey` on the other kindles with keyboards and `hasScreenKB` on K4.

**Comment by Commodore64user** [comment]
You are also ignoring #11864 which fixes the fact that content selection does not actually work atm.

**Comment by Frenzie** [comment]
@Commodore64user implemented as per your suggestion

File: frontend/apps/reader/modules/readerlink.lua [hunk 1]
```diff
@@ -233,15 +233,25 @@ end
...
 function ReaderLink:registerKeyEvents()
-    if Device:hasKeys() and not Device:hasFiveWay() then
+    if Device:hasScreenKB() or Device:hasSymKey() then
+        self.key_events.GotoSelectedPageLink = { { "Press" }, event = "GotoSelectedPageLink" }
+        if Device:hasKeyboard() then
+            self.key_events.AddCurrentLocationToStack = { { "Shift", "Down" } }
+            self.key_events.SelectNextPageLink = { { "Shift", "LPgFwd" }, event = "SelectNextPageLink" }
+            self.key_events.SelectPrevPageLink = { { "Shift", "LPgBack" }, event = "SelectPrevPageLink" }
+        else
+            self.key_events.AddCurrentLocationToStack = { { "ScreenKB", "Down" } }
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerlink.lua:243]: could you change this to `ScreenKB` and `Press` (instead of `Down`) there is something I want to try and if successful would go on KB+Down. Also, any idea why this doesn't show a notification when one uses it? normally at the top of the screen a little box appears saying, "added to stack or whatever" (yes all notifications under Screen menu are ON).

File: frontend/apps/reader/modules/readerlink.lua [hunk 2]
```diff
@@ -233,15 +233,25 @@ end
 function ReaderLink:onGesture() end
 
 function ReaderLink:registerKeyEvents()
-    if Device:hasKeys() and not Device:hasFiveWay() then
+    if Device:hasScreenKB() or Device:hasSymKey() then
+        self.key_events.GotoSelectedPageLink = { { "Press" }, event = "GotoSelectedPageLink" }
+        if Device:hasKeyboard() then
+            self.key_events.AddCurrentLocationToStack = { { "Shift", "Down" } }
```
Comment (Commodore64user) [inline frontend/apps/reader/modules/readerlink.lua:239]: same here `Shift` + `Press`

File: frontend/apps/reader/modules/readerlink.lua (hunk 1 above)
Comment (Frenzie) [reply frontend/apps/reader/modules/readerlink.lua:243]: It looks like notification true or false is passed as an argument. DIspatcher sets arg=true.

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerlink.lua:243]: I’ve already tried that and still won’t show it… not sure why

Comment (Frenzie) [reply frontend/apps/reader/modules/readerlink.lua:243]: Hm yes, the true comes through if you add `args=true` but it doesn't seem to show the notification. You say it works from Dispatcher?

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerlink.lua:243]: Yes.

Comment (Frenzie) [reply frontend/apps/reader/modules/readerlink.lua:243]: I think @NiLuJe last touched notifications. I don't know much about them otoh.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerlink.lua:243]: Because it's `args`, plural, for `InputContainer`'s `key_events` ;).

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerlink.lua:243]: It's the same both with arg or args, i.e does not work

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerlink.lua:243]: Right, take two then ;D.

IIRC, Dispatcher brackets its calls in a way that ensure the notification source is set as Dispatcher.

Here, that's not the case, so the defaults (or the last set source if there's some misbehaving code) is used, and it's likely to be masked out in the default settings, so the notification doesn't show.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerlink.lua:243]: > the defaults

Being `SOURCE_OTHER`, which is described as `all other sources (e.g. keyboard)`, which is indeed masked out unless you explicitly request notifications from *all* sources.

TL;DR: behaving as expected ;).

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerlink.lua:243]: Right, so how do we set a source and/or make it show the notification?

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerlink.lua:243]: We... don't because it's behaving as expected and keyboard shortcuts shouldn't display notification unless the user asks for it?

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerlink.lua:243]: Yes, but this is 'different'...? I need the notification to be shown so it provides feedback on NT kindles, otherwise it appears as if nothing was happening when that would just not the case.

Essentially it should work same as if it was coming from dispatcher.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerlink.lua:243]: Then you need a custom event that either wraps the current one in `Notify:setNotifySource(NOTIFICATION.SOURCE_ALWAYS_SHOW)` <-> `Notification:resetNotifySource()` or reimplement the current one except that it passes `NOTIFICATION.SOURCE_ALWAYS_SHOW` as the second arg to `Notify`.

Comment (Commodore64user) [reply frontend/apps/reader/modules/readerlink.lua:243]: Slightly unrelated question, do you have this SOED dictionary @NiLuJe? (mentioned here https://www.mobileread.com/forums/showthread.php?t=347490)

It would be a shame if someone were to accidentally upload it here for 12hrs tops, wink. And then deleted it for ever once they realised their mistake.
